GP11 : AE functions fix/panic reason

cedric-chaumont-st-dev opened the pull request:
Signed-off-by: Cedric Chaumont <cedric.chaumont@st.com>

pascal-brand-st-dev commented:
Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

cedric-chaumont-st-dev commented:
Rebased.

jenswi-linaro commented on lib/libutee/tee_api_operations.c:
```diff
@@ -1224,46 +1273,69 @@ TEE_Result TEE_AEDecryptFinal(TEE_OperationHandle op,
-	if (tagLen != op->ae_tag_len)
+	if (tagLen != operation->ae_tag_len) {
 		res = TEE_ERROR_MAC_INVALID;
+		goto out;
```
Isn't the operation to be considered finalized (`TEE_HANDLE_FLAG_INITIALIZED` to be cleared) here also?

cedric-chaumont-st-dev replied:
@jenswi-linaro :  The spec is not really clear about this. Why do you think it is better to consider the operation finalized in that case (same as previous impletementation) since we have an error message (compared to other error messages such as TEE_ERROR_SHORT_BUFFER) ?

jenswi-linaro replied:
For `TEE_ERROR_SHORT_BUFFER` it would make sense to call the function again with a larger buffer. With `TEE_ERROR_MAC_INVALID` the data has already been consumed, what more could the caller do with the the operation handle?

cedric-chaumont-st-dev replied:
I would prefer to keep it initialized as long as long as an operation is not successfully complete . I think the caller should handle the error messages by calling either TEE_ResetOperation  or TEE_FreeOperation.

jenswi-linaro replied:
The operation is complete, there's nothing left to do with the operation handle in that state (how can the application know if the tag was truly invalid or just a tagLen mismatch?). It would be an error of the application to call this function again in that state.

*The operation may be in either initial or active state and enters initial state afterwards.*

What isn't clear about that is what to do when returning `TEE_ERROR_SHORT_BUFFER`, that error code doesn't make sense unless we allow the function to be called again.

cedric-chaumont-st-dev commented:
Rebased. Tested (STM ARM Juno).
28701 subtests of which 0 failed
687 test cases of which 0 failed
0 test case was skipped
TEE test application done!

jenswi-linaro commented:
```
Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>
```

cedric-chaumont-st-dev commented:
Rebased. Squashed. Tested (STM, ARM Juno).
28701 subtests of which 0 failed
687 test cases of which 0 failed
0 test case was skipped
TEE test application done!
